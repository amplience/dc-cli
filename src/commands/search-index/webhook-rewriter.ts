function matchAll(regex: RegExp, string: string): RegExpExecArray[] {
  const result: RegExpExecArray[] = [];
  let match: RegExpExecArray | null;

  while ((match = regex.exec(string)) != null) {
    result.push(match);
  }

  return result;
}

export function rewriteDeliveryContentItem(
  webhookBody: string,
  account: string,
  stagingEnvironment: string | undefined
): string {
  // Current limitations - cannot handle key/value pairs where the value contains space, or a }} within quotes.
  // These should not affect the two variables being replaced anyways.
  // First, locate the withDeliveryContentItem opening tags.

  const tagRegexG = /\{\{\#withDeliveryContentItem(?<whitespace>\s+)(?<body>.+\=.+(\s+.+\=.+)*\s*)\}\}/g;
  const keyValueRegex = /(?<key>.+?)\=(?<value>.+?)\s+/g;

  const tagMatches = matchAll(tagRegexG, webhookBody);

  let replaceOffset = 0;

  for (const tag of tagMatches) {
    const tGroups = tag.groups as { [key: string]: string };

    const matchIndex = tag.index;
    const body = tGroups.body + ' ';
    const bodyIndex = matchIndex + '{{#withDeliveryContentItem'.length + tGroups.whitespace.length;

    const keyValueMatches = matchAll(keyValueRegex, body);

    for (const pair of keyValueMatches) {
      const pGroups = pair.groups as { [key: string]: string };

      const key = pGroups.key;
      const keyIndex = bodyIndex + pair.index + replaceOffset;

      const value = pGroups.value;
      const valueIndex = keyIndex + key.length + 1;

      let replaceValue: string | null = null;

      switch (key) {
        case 'account':
          replaceValue = account;
          break;
        case 'stagingEnvironment':
          if (stagingEnvironment !== undefined) {
            replaceValue = stagingEnvironment;
          }
          break;
      }

      if (replaceValue != null) {
        replaceValue = `"${replaceValue.replace(/\\/g, '\\\\').replace(/"/g, '"')}"`;
        webhookBody = webhookBody.substr(0, valueIndex) + replaceValue + webhookBody.substr(valueIndex + value.length);

        replaceOffset += replaceValue.length - value.length;
      }
    }
  }

  return webhookBody;
}
