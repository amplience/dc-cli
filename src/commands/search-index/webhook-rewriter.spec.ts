import { rewriteDeliveryContentItem } from './webhook-rewriter';

describe('webhook-rewriter tests', function() {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('should not rewrite an empty string', async () => {
    expect(rewriteDeliveryContentItem('', 'account', 'staging')).toEqual('');
  });

  it('should not rewrite a simple string', async () => {
    expect(rewriteDeliveryContentItem('basic', 'account', 'staging')).toEqual('basic');
  });

  it('should not rewrite any unrelated tag', async () => {
    expect(rewriteDeliveryContentItem('{{#unrelated}}{{/unrelated}}', 'account', 'staging')).toEqual(
      '{{#unrelated}}{{/unrelated}}'
    );
  });

  it('should rewrite a simple webhook', async () => {
    const example = `{
{{#withDeliveryContentItem contentItemId=payload.id account="tobereplaced" stagingEnvironment="old"}}
  "content" : "{{~#each (first (pluck contentBlocks "content") 4)~}}{{~#each this.values~}}{{#if (eq locale "en-GB")}}{{{truncate value 1000}}}{{/if}}{{~/each~}}{{#unless @last}};{{/unless~}}{{~/each~}}"
{{/withDeliveryContentItem}}
}`;
    expect(rewriteDeliveryContentItem(example, 'account', 'staging')).toMatchSnapshot();
  });

  it('should rewrite multiple tags within the same webhook', async () => {
    const example = `{
{{#withDeliveryContentItem contentItemId=payload.id account="tobereplaced" stagingEnvironment="old"}}
  {{example1}}
{{/withDeliveryContentItem}}
  "key": "value",
{{#withDeliveryContentItem contentItemId=payload.id account="tobereplaced2" stagingEnvironment="old2"}}
  {{example1}}
{{/withDeliveryContentItem}}
}`;
    expect(rewriteDeliveryContentItem(example, 'account', 'staging')).toMatchSnapshot();
  });

  it('should rewrite only one argument if the other is not present', async () => {
    const example = `{
{{#withDeliveryContentItem contentItemId=payload.id account="tobereplacednostaging"}}
  {{example1}}
{{/withDeliveryContentItem}}
}`;
    expect(rewriteDeliveryContentItem(example, 'account', "doesn't appear")).toMatchSnapshot();
  });

  it('should not rewrite a tag with no matching arguments', async () => {
    const example = `{
{{#withDeliveryContentItem contentItemId=payload.id"}}
  {{example1}}
{{/withDeliveryContentItem}}
}`;
    expect(rewriteDeliveryContentItem(example, 'account', "doesn't appear")).toEqual(example);
  });

  it('should escape quotes and backslashes', async () => {
    const example = `{
{{#withDeliveryContentItem contentItemId=payload.id account="1" stagingEnvironment="2"}}
  {{example1}}
{{/withDeliveryContentItem}}
}`;
    expect(rewriteDeliveryContentItem(example, '"quotedString\\with\\backslash"', 'back\\slash')).toMatchSnapshot();
  });
  it('should still replace values with unusual whitespace', async () => {
    const example = `{
{{#withDeliveryContentItem 
    contentItemId=payload.id 
    account="1" 
    stagingEnvironment="another"\t    }}
  {{example1}}
{{/withDeliveryContentItem}}
}`;
    expect(rewriteDeliveryContentItem(example, 'replaced', 'whitespace')).toMatchSnapshot();
  });
});
