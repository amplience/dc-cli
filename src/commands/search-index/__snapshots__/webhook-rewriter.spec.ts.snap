// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webhook-rewriter tests should escape quotes and backslashes 1`] = `
"{
{{#withDeliveryContentItem contentItemId=payload.id account=\\"\\"quotedString\\\\\\\\with\\\\\\\\backslash\\"\\" stagingEnvironment=\\"back\\\\\\\\slash\\"}}
  {{example1}}
{{/withDeliveryContentItem}}
}"
`;

exports[`webhook-rewriter tests should rewrite a simple webhook 1`] = `
"{
{{#withDeliveryContentItem contentItemId=payload.id account=\\"account\\" stagingEnvironment=\\"staging\\"}}
  \\"content\\" : \\"{{~#each (first (pluck contentBlocks \\"content\\") 4)~}}{{~#each this.values~}}{{#if (eq locale \\"en-GB\\")}}{{{truncate value 1000}}}{{/if}}{{~/each~}}{{#unless @last}};{{/unless~}}{{~/each~}}\\"
{{/withDeliveryContentItem}}
}"
`;

exports[`webhook-rewriter tests should rewrite multiple tags within the same webhook 1`] = `
"{
{{#withDeliveryContentItem contentItemId=payload.id account=\\"account\\" stagingEnvironment=\\"staging\\"}}
  {{example1}}
{{/withDeliveryContentItem}}
  \\"key\\": \\"value\\",
{{#withDeliveryContentItem contentItemId=payload.id account=\\"account\\" stagingEnvironment=\\"staging\\"}}
  {{example1}}
{{/withDeliveryContentItem}}
}"
`;

exports[`webhook-rewriter tests should rewrite only one argument if the other is not present 1`] = `
"{
{{#withDeliveryContentItem contentItemId=payload.id account=\\"account\\"}}
  {{example1}}
{{/withDeliveryContentItem}}
}"
`;

exports[`webhook-rewriter tests should still replace values with unusual whitespace 1`] = `
"{
{{#withDeliveryContentItem 
    contentItemId=payload.id 
    account=\\"replaced\\" 
    stagingEnvironment=\\"whitespace\\"	    }}
  {{example1}}
{{/withDeliveryContentItem}}
}"
`;
